build/generate-tray-icons.ts

```diff
@@ -3,48 +3,128 @@
-const lightTemplate = svgRoot.outerHTML;
+function getSvgStyling(colouring: "dark" | "light"): string {
+  return `
+    <style>
+      ellipse {
+        stroke: ${colouring === "dark" ? "white" : "black"} !important;
+      }
+      path, rect {
+        fill: ${colouring === "dark" ? "white" : "black"} !important;
+      }
+    </style>
+  `;
+}
 
-svgRoot.innerHTML += `<style>* {fill: black !important;}</style>`;
+type TargetSystems = "macos" | "windows-or-linux";
 
-const darkTemplate = svgRoot.outerHTML;
+async function getBaseIconImage(system: TargetSystems) {
+  const svgData = await readFile(inputFile, { encoding: "utf-8" });
+  const dom = new JSDOM(`<body>${svgData}</body>`);
+  const root = dom.window.document.body.getElementsByTagName("svg")[0];
 
-console.log("Generating tray icon pngs");
+  root.innerHTML += getSvgStyling(system === "macos" ? "light" : "dark");
 
-ensureDirSync(outputFolder);
+  return Buffer.from(root.outerHTML);
+}
 
-Promise.all([
-  sharp(Buffer.from(lightTemplate))
+async function generateImage(image: Buffer, size: number, namePrefix: string) {
+  sharp(image)
     .resize({ width: size, height: size })
     .png()
-    .toFile(path.join(outputFolder, "trayIconDarkTemplate.png")),
-  sharp(Buffer.from(lightTemplate))
-    .resize({ width: size*2, height: size*2 })
-    .png()
-    .toFile(path.join(outputFolder, "trayIconDarkTemplate@2x.png")),
-  sharp(Buffer.from(darkTemplate))
-    .resize({ width: size, height: size })
-    .png()
-    .toFile(path.join(outputFolder, "trayIconTemplate.png")),
-  sharp(Buffer.from(darkTemplate))
-    .resize({ width: size*2, height: size*2 })
-    .png()
-    .toFile(path.join(outputFolder, "trayIconTemplate@2x.png")),
-])
-  .then((resolutions) => console.log(`Generated ${resolutions.length} images`))
-  .catch(console.error);
+    .toFile(path.join(outputFolder, `${namePrefix}.png`));
+}
+
+async function generateImages(image: Buffer, size: number, name: string) {
+  await Promise.all([
+    generateImage(image, size, name),
+    generateImage(image, size*2, `${name}@2x`),
+    generateImage(image, size*3, `${name}@3x`),
+    generateImage(image, size*4, `${name}@4x`),
+  ]);
+}
+
+async function generateUpdateAvailableImages(baseImage: Buffer, system: TargetSystems) {
+  const noticeIconImage = await getNoticeIconImage(system);
+  const circleBuffer = await sharp(Buffer.from(`
+    <svg viewBox="0 0 64 64">
+      <circle cx="32" cy="32" r="32" fill="black" />
+    </svg>
+  `))
+    .toBuffer();
+
+  return sharp(baseImage)
+    .resize({ width: 128, height: 128 })
+    .composite([
+      {
+        input: circleBuffer,
+        top: 64,
+        left: 64,
+        blend: "dest-out",
+      },
+      {
+        input: (
+          await sharp(noticeIconImage)
+            .resize({
+              width: 60,
+              height: 60,
+            })
+            .toBuffer()
+        ),
+        top: 66,
+        left: 66,
+      },
+    ])
+    .toBuffer();
+}
```

Comment:
1. How this code can do the different things without modifications? If not -> you still need to edit something and most easy way to update image it's just replace it
2. I'm not agains generating different sizes programatically :) only agains composition from different files (you're using positioning, etc)

Reply:
I do agree that the simplest solution would be that we have set of icons already generated in the repositories. We could make it so that icons could be configured in `package.json` so that we can have different icons in lens-ide. In that way we are not tied to having same file names or even directory structure in lens-ide and it would make it easier to understand what happens. (package.json -files are merged and then used as a configuration)

```json

{
  "icons": {
    "tray": {
      "normal": "./icons/tray/normal-icon.svg",
      "normal-2x": "./icons/tray/normal-icon@2x.svg",
      "update-available": "./icons/tray/update-available-icon.svg"
   }
  }
}

```

However, having automation that creates the icons is good. Manually editing icons leads to place where changing the icon is hard and can be done only by someone who has correct tools for the job.

We could have best of both worlds; script that generates icons (NPM-script, which is only triggered manually) but still have all generated icons in the repositories. 

---
This is bit against Open Closed Principle and we could make it comply with it, the simplicity is the winner.

Reply:
But, needs to be said that this is already an improvement. So I don't see this as critical.